src/analysis/convex/caratheodory.lean

```diff
@@ -0,0 +1,242 @@
+      apply Union_subset_Union_const,
+      exact λ h, set.subset.trans h (erase_subset _ _), } }
+end
+
+/--
+The convex hull of any finset `t` is contained in
+the union of the convex hulls of the finsets `t' ⊆ t` with `t'.card ≤ findim ℝ E + 1`.
+-/
+lemma shrink (t : finset E) :
+  convex_hull (↑t : set E) ⊆
+    ⋃ (t' : finset E) (w : t' ⊆ t) (b : t'.card ≤ findim ℝ E + 1), convex_hull ↑t' :=
+begin
+  by_cases h : t.card ≤ findim ℝ E + 1,
+  { apply subset_subset_Union t,
+    apply subset_subset_Union (set.subset.refl _),
+    exact subset_subset_Union h (set.subset.refl _), },
+  push_neg at h,
+  obtain ⟨k, w⟩ := le_iff_exists_add.mp (le_of_lt h), clear h,
+  exact shrink' _ _ w,
+end
+
+end caratheodory
+
+/--
+One inclusion of Carathéodory's convexity theorem.
+
+The convex hull of a set `s` in ℝᵈ is contained in
+the union of the convex hulls of the (d+1)-tuples in `s`.
+-/
+lemma convex_hull_subset_union (s : set E) :
+  convex_hull s ⊆ ⋃ (t : finset E) (w : ↑t ⊆ s) (b : t.card ≤ findim ℝ E + 1), convex_hull ↑t :=
+begin
+  -- First we replace `convex_hull s` with the union of the convex hulls of finite subsets,
+  rw convex_hull_eq_union_convex_hull_finite_subsets,
+  -- and prove the inclusion for each of those.
+  apply Union_subset, intro r,
+  apply Union_subset, intro h,
+  -- Second, for each convex hull of a finite subset, we shrink it.
+  transitivity,
+  { apply caratheodory.shrink, },
+  { -- After that it's just shuffling unions around.
+    apply Union_subset, intro t,
+    apply Union_subset, intro htr,
+    apply Union_subset, intro ht,
+    apply set.subset_subset_Union t,
+    apply set.subset_subset_Union (subset.trans htr h),
+    exact subset_Union _ ht, },
+end
+
+/--
+Carathéodory's convexity theorem.
+
+The convex hull of a set `s` in ℝᵈ is the union of the convex hulls of the (d+1)-tuples in `s`.
+-/
+theorem convex_hull_eq_union (s : set E) :
+  convex_hull s = ⋃ (t : finset E) (w : ↑t ⊆ s) (b : t.card ≤ findim ℝ E + 1), convex_hull ↑t :=
+begin
+  apply set.subset.antisymm,
+  { apply convex_hull_subset_union, },
+  iterate 3 { convert Union_subset _, intro, },
+  exact convex_hull_mono ‹_›,
+end
+
+/--
+A more explicit formulation of Carathéodory's convexity theorem,
+writing an element of a convex hull as the center of mass
+of an explicit `finset` with cardinality at most `dim + 1`.
+-/
+theorem eq_center_mass_card_le_dim_succ_of_mem_convex_hull
+  {s : set E} {x : E} (h : x ∈ convex_hull s) :
+  ∃ (t : finset E) (w : ↑t ⊆ s) (b : t.card ≤ findim ℝ E + 1)
+    (f : E → ℝ), (∀ y ∈ t, 0 ≤ f y) ∧ t.sum f = 1 ∧ t.center_mass f id = x :=
+begin
+  rw convex_hull_eq_union at h,
+  simp only [exists_prop, mem_Union] at h,
+  obtain ⟨t, w, b, m⟩ := h,
+  refine ⟨t, w, b, _⟩,
+  rw finset.convex_hull_eq at m,
+  simpa only [exists_prop] using m,
+end
+
+/--
+A slight variation of `eq_center_mass_card_le_dim_succ_of_mem_convex_hull`,
+asserting that the coefficients can be taken to be strictly positive.
```

Comment:
This doc string should be more explicit.

Reply:
Done!